### DISCUSSION TITLE
Custom shader using vertex coloring broken from 0.37.0->0.38.1

### DISCUSSION THREAD

**Discussion by rileychall** [pyvista/pyvista#5200] [Q&A] [answered]
I'm using a custom shader that makes use of the vertex coloring in the VTK PolyData shaders. VTK constructs those shaders at runtime based on the properties of the mapper. Vertex coloring is added to the shader in [this section of `vtkOpenGLPolyDataMapper`](https://gitlab.kitware.com/vtk/vtk/-/blob/master/Rendering/OpenGL2/vtkOpenGLPolyDataMapper.cxx#L742), where `out vec4 vertexColorVSOutput;` is declared.

The problem is that the vertex coloring is added in PyVista 0.37.0 and it is not added in 0.38.1 and beyond. The constructed shaders do not contain `vertexColorVSOutput`. This is while running identical code and on two Conda environments that differ by only the PyVista version.
(I imagine vertex coloring involves more than just adding that line to the shader, but that is the change that I'm actually able to observe.)

My two main suspects for why this is happening are the addition of `show_vertices` (#3745) and that it appears that `vtkPolyData` objects are being wrapped into `pyvista.PolyData` where they weren't in 0.37.0. But I haven't been able to figure out how those might prevent vertex coloring from being enabled.

I'm doing some very hacky, very custom stuff, so this is probably a long shot. It's also deep inside a Qt application so making a minimal example would take a fair bit of work.

Any help or insight would be hugely appreciated!

**Comment by rileychall** [comment]
Here is a minimal example, which works as desired with 0.37.0 and throws an error with 0.38.1 and 0.42.3:
<details>
  <summary><b>Example code</b></summary>

  ```python
  import numpy as np
  import pyvista as pv
  import vtk
  
  
  def main():
      vertices = np.array([[0, 0, 0], [1, 0, 0], [1, 1, 0], [0, 1, 0], [0.5, 0.5, -1]])
      pd = pv.PolyData(vertices)
      pd["values"] = np.array(range(5))
      algo = vtk.vtkVertexGlyphFilter()
      algo.SetInputData(pd)
      algo.Update()
      mapper = pv.DataSetMapper()
      mapper.SetInputConnection(algo.GetOutputPort())
      actor = pv.Actor(mapper=mapper)
      actor.GetProperty().SetPointSize(30)
  
      # Unmodified
      plotter1 = pv.Plotter()
      plotter1.add_actor(actor, name="main")
      plotter1.show()
  
      # Modified shaders
      plotter2 = pv.Plotter()
      actor2, _ = plotter2.add_actor(actor, name="main")
  
      shaders = actor2.GetShaderProperty()
      shaders.SetGeometryShaderCode(minimal_GS)
  
      plotter2.show()
  
  
  minimal_GS = """
  //VTK::System::Dec
  
  //VTK::Color::Dec
  //// when replaced by VTK, this becomes:
  // in vec4 vertexColorVSOutput[];
  // out vec4 vertexColorGSOutput;
  
  layout(points) in;
  layout(points) out;
  
  void main()
  {
      vertexColorGSOutput = vertexColorVSOutput[0] + gl_in[0].gl_Position;
      gl_Position = gl_in[0].gl_Position;
      EmitVertex();
  }
  """
  
  if __name__ == "__main__":
      main()
  ```
</details>

The error with 0.38.1 is caused by `vertexColorVSOutput` and `vertexColorGSOutput` not being declared in the shader, because VTK didn't enable vertex coloring in the geometry shader. The vertex color activation thing only happens in the geometry shader in two places in VTK:
1. [`vtkOpenGLPolyDataMapper`](https://gitlab.kitware.com/vtk/vtk/-/blob/master/Rendering/OpenGL2/vtkOpenGLPolyDataMapper.cxx#L742) (same one as above)
2. [`vtkOpenGLGlyph3DHelper`](https://gitlab.kitware.com/vtk/vtk/-/blob/master/Rendering/OpenGL2/vtkOpenGLGlyph3DHelper.cxx#L111)

I'm not sure which one is actually in play here.

**Reply by MatthewFlamm** [reply]
Thanks for running this down.  I'm not totally familiar with the workings of this part of the code.  But for subtle changes, it might help to do a `git bisect` to pinpoint the commit that causes the issue you are seeing.  To me, it seems like #3745 only adds some convenience options.

**Reply by rileychall** [reply]
That's a good call on doing a bisect. It identified PR #3870 as the first bad commit. Bisecting that PR narrowed it down to [this commit](https://github.com/pyvista/pyvista/pull/3870/commits/13b4a73ec8b7a6f79cdfd4390f42c6a8d0c1e39f). I'm looking more closely now.

**Comment by rileychall** [comment, marked as answer]
Figured it out! It's because `interpolate_before_map` was changed to default to True. Adding `mapper.interpolate_before_map = False` to the example script causes it to work correctly with both 0.37.0 and 0.42.3.

**Reply by MatthewFlamm** [reply]
`git bisect` to the rescue!  Thanks for coming back here.
